From f08b19227bb9a031b1f63b7452f7e50ff53f3700 Mon Sep 17 00:00:00 2001
From: "H. Vetinari" <h.vetinari@gmx.com>
Date: Sat, 29 Oct 2022 09:39:17 +1100
Subject: [PATCH 2/4] use Ninja everywhere

---
 setup.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index 49a847c..e4b168c 100644
--- a/setup.py
+++ b/setup.py
@@ -36,7 +36,7 @@ class get_pybind_include(object):
 cmake_args = []
 
 # Add parameters to cmake_args and define_macros
-cmake_args = ["-DUNITTESTS=OFF"]
+cmake_args = ["-DUNITTESTS=OFF", "-GNinja"]
 cmake_build_flags = []
 lib_subdir = []
 
-- 
2.43.0.windows.1

