From 450afd6e65e6740454ac8c4d0d6cf8be40114ea3 Mon Sep 17 00:00:00 2001
From: "H. Vetinari" <h.vetinari@gmx.com>
Date: Sat, 29 Oct 2022 11:16:16 +1100
Subject: [PATCH 4/4] build qdldlamd as shared lib and install it

---
 c/CMakeLists.txt |  8 +++++++-
 setup.py         | 26 +++++++++++++++-----------
 2 files changed, 22 insertions(+), 12 deletions(-)

diff --git a/c/CMakeLists.txt b/c/CMakeLists.txt
index 33cd12e..000c6a5 100644
--- a/c/CMakeLists.txt
+++ b/c/CMakeLists.txt
@@ -31,8 +31,14 @@ set(
 
 
 # Add QDLDL+AMD library
-add_library(qdldlamd STATIC ${amd_src})
+add_library(qdldlamd SHARED ${amd_src})
 target_link_libraries(qdldlamd qdldl)
 target_include_directories(qdldlamd PRIVATE
 	                   ${CMAKE_CURRENT_SOURCE_DIR}/amd/include
 			    )
+
+install(TARGETS qdldlamd
+    RUNTIME DESTINATION "bin"
+    LIBRARY DESTINATION "lib"
+    ARCHIVE DESTINATION "lib"
+)
diff --git a/setup.py b/setup.py
index 8b06645..bc04e3d 100644
--- a/setup.py
+++ b/setup.py
@@ -38,18 +38,23 @@ prefix = os.environ["LIBRARY_PREFIX"] if system() == 'Windows' else os.environ["
 cmake_args = []
 
 # Add parameters to cmake_args and define_macros
-cmake_args = ["-DUNITTESTS=OFF", "-GNinja"]
-cmake_build_flags = []
+cmake_args = ["-DUNITTESTS=OFF", "-GNinja", f'-DCMAKE_INSTALL_PREFIX={prefix}']
+cmake_build_flags = ['--config', 'Release']
 lib_subdir = []
 
 # Check if windows linux or mac to pass flag
 if system() == 'Windows':
-    cmake_build_flags += ['--config', 'Release']
-    lib_name = 'qdldlamd.lib'
-    lib_subdir = ['Release']
+    cmake_args += ['-DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=ON']
+    lib_prefix = ''
+    lib_ext = '.lib'
+elif system() == "Linux":
+    lib_prefix = 'lib'
+    lib_ext = '.so'
+else:
+    lib_prefix = 'lib'
+    lib_ext = '.dylib'
 
-else:  # Linux or Mac
-    lib_name = 'libqdldlamd.a'
+lib_name = lib_prefix + "qdldl"
 
 # Set optimizer flag
 if system() != 'Windows':
@@ -61,9 +66,6 @@ else:
 current_dir = os.getcwd()
 qdldl_dir = os.path.join(current_dir, 'c',)
 qdldl_build_dir = os.path.join(qdldl_dir, 'build')
-qdldl_lib = [qdldl_build_dir, 'out'] + lib_subdir + [lib_name]
-qdldl_lib = os.path.join(*qdldl_lib)
-
 
 class build_ext_qdldl(build_ext):
     def build_extensions(self):
@@ -82,6 +84,7 @@ class build_ext_qdldl(build_ext):
         call(['cmake'] + cmake_args + ['..'])
         call(['cmake', '--build', '.', '--target', 'qdldlamd'] +
              cmake_build_flags)
+        call(['cmake', '--install', '.'])
 
         # Change directory back to the python interface
         os.chdir(current_dir)
@@ -98,7 +101,8 @@ qdldl = Extension('qdldl',
                                 get_pybind_include(user=False)],
                   language='c++',
                   extra_compile_args=compile_args + ['-std=c++11'],
-                  extra_objects=[qdldl_lib])
+                  extra_objects=[os.path.join(prefix, "lib", x + lib_ext)
+                                 for x in [lib_name, lib_name + "amd"]])
 
 
 def readme():
-- 
2.43.0.windows.1

